UCP/WIREUP/IB: Fix error message when FLID is not available

yosefe opened the pull request:
Before:

`[1740477778.122527] [swx-ucx-ibr-03:1950035:0]          select.c:641  UCX  ERROR   no active messages transport to <no debug data>: self/memory - Destination is unreachable, rc_verbs/mlx5_0:1 - Destination is unreachable`


After:

`[1740487396.304113] [swx-ucx-ibr-03:1951997:0]          select.c:642  UCX  ERROR   no active messages transport to swx-ucx-ibr-01:3618365: self/memory - unreachable, rc_verbs/mlx5_0:1 - different subnet prefix 0xfec0000000000002/0xfec0000000000001 and FLID is not available`

brminich commented on src/uct/ib/base/ib_iface.c:
```diff
@@ -883,22 +878,22 @@ static int uct_ib_iface_dev_addr_is_reachable(
-            return 0;
+        if (!iface->config.flid_enabled) {
+            flid_info_str = "disabled";
+        } else if ((uct_ib_iface_gid_extract_flid(&iface->gid_info.gid)) == 0 ||
```
```suggestion
        } else if ((uct_ib_iface_gid_extract_flid(&iface->gid_info.gid) == 0) ||
```

brminich commented on src/ucp/wireup/select.c:
```diff
@@ -623,9 +623,8 @@ static UCS_F_NOINLINE ucs_status_t ucp_wireup_select_transport(
-            snprintf(p, endp - p, UCT_TL_RESOURCE_DESC_FMT" - %s, ",
-                     UCT_TL_RESOURCE_DESC_ARG(resource),
-                     ucs_status_string(UCS_ERR_UNREACHABLE));
+            snprintf(p, endp - p, UCT_TL_RESOURCE_DESC_FMT " - %s, ",
```
can use `ucs_snprintf_safe`

brminich commented on src/ucp/wireup/select.c:
```diff
@@ -623,9 +623,8 @@ static UCS_F_NOINLINE ucs_status_t ucp_wireup_select_transport(
-                     UCT_TL_RESOURCE_DESC_ARG(resource),
-                     ucs_status_string(UCS_ERR_UNREACHABLE));
+            snprintf(p, endp - p, UCT_TL_RESOURCE_DESC_FMT " - %s, ",
+                     UCT_TL_RESOURCE_DESC_ARG(resource), info_str);
```
so it also fixes printing other unreachable reasons, right?

brminich commented on src/uct/ib/base/ib_iface.c:
```diff
@@ -841,19 +841,14 @@ static int uct_ib_iface_gid_extract_flid(const union ibv_gid *gid)
-static int uct_ib_iface_is_flid_enabled(uct_ib_iface_t *iface)
-{
-    return iface->config.flid_enabled &&
-           (uct_ib_iface_gid_extract_flid(&iface->gid_info.gid) != 0);
```
why is this check removed?

yosefe replied:
uct_ib_iface_is_flid_enabled is not needed anymore since in uct_ib_iface_dev_addr_is_reachable the code was refactored to print the reason, and in uct_ib_iface_resolve_remote_flid() function uct_ib_iface_gid_extract_flid is already checked in the end.

brminich replied:
`uct_ib_iface_resolve_remote_flid` checks for flid in the remote gid, but you removed check for local gid in iface

yosefe replied:
fixed

brminich commented on src/ucp/wireup/select.c:
```diff
@@ -592,7 +592,7 @@ static UCS_F_NOINLINE ucs_status_t ucp_wireup_select_transport(
 
         is_reachable = 0;
-
+        ucs_snprintf_safe(info_str, info_str_size, "unreachable");
```
maybe use snprintf here for consistency then? In this function only `snprintf` is used

yosefe replied:
fixed

amastbaum commented:
@yosefe do you think we still need to be able to print an info string from `ucp_wireup_select_lanes`? What you added makes pretty much the same thing as intended in `select.c:2513` but better.

* we may consider removing the code that resets `info_str_buf` in `uct_iface_is_reachable_v2` as the info buffer is filled with "unreachable" before this function is called.

amastbaum commented on src/ucp/wireup/select.c:
```diff
@@ -592,7 +592,7 @@ static UCS_F_NOINLINE ucs_status_t ucp_wireup_select_transport(
 
         is_reachable = 0;
-
+        ucs_snprintf_safe(info_str, info_str_size, "unreachable");
```
AFAIR we currently enable `info_str` to be NULL and its value is checked inside `uct_iface_fill_info_str_buf`. I actually don't remember why we didn't pass an info str buffer to all `.._select_lanes` functions, but maybe now it can be removed (as I mentioned in the general comment) and `info_str' can be defined inside `ucp_wireup_select_transport`.. wdyt?

yosefe replied:
removed info_str from parameters and moved it inside select_transport

yosefe commented:
> @yosefe do you think we still need to be able to print an info string from `ucp_wireup_select_lanes`? What you added makes pretty much the same thing as intended in `select.c:2513` but better.
> 

Removed it from select.c:2513

> * we may consider removing the code that resets `info_str_buf` in `uct_iface_is_reachable_v2` as the info buffer is filled with "unreachable" before this function is called.

We still need to reset uct_info string to "not available" in UCP before going over remote addresses, since ucp_wireup_is_reachable() can return 0 also if tl name is different - without calling uct_iface_is_reachable_v2.

Added a gtest that UCT is_reachable always fills the info string

brminich commented on src/uct/ib/mlx5/rc/rc_mlx5_iface.c:
```diff
@@ -668,10 +678,11 @@ uct_rc_mlx5_iface_is_reachable_v2(const uct_iface_h tl_iface,
-                    params, "incompatible hardware tag matching. "
-                    "%s (local) vs %s (remote)",
-                    tm_type_to_str[my_type], tm_type_to_str[remote_type]);
+        uct_iface_fill_info_str_buf(params,
+                                    "incompatible hardware tag matching. "
+                                    "%s (local) vs %s (remote)",
+                                    uct_rc_mlx5_iface_tm_type_str(my_type),
+                                    uct_rc_mlx5_iface_tm_type_str(remote_type));
```
seems to be unrelated

yosefe replied:
it is needed because the is_reachable test (with garbage address) causes access to invalid memory without this fix

brminich reviewed (commented):
test errors seem to be relevant

yosefe commented:
@brminich can you pls take a look?

brminich commented on src/uct/ib/base/ib_iface.h:
```diff
@@ -526,8 +526,8 @@ void uct_ib_iface_address_pack(uct_ib_iface_t *iface, uct_ib_address_t *ib_addr)
  */
-void uct_ib_address_unpack(const uct_ib_address_t *ib_addr,
-                           uct_ib_address_pack_params_t *params_p);
+ucs_status_t uct_ib_address_unpack(const uct_ib_address_t *ib_addr,
```
minor: return status is not mentioned in the doxygen comment above

brminich replied:
it is unresolved (seems disappeared after some force push)

yosefe commented:
force-pushed to revert to [6318ce0](https://github.com/openucx/ucx/commit/6318ce07ac5f093778cb5556036d0263640aec31) + few code style fixes
https://github.com/openucx/ucx/compare/6318ce07a..87cf6128721508a940796eeb964e186544d2a371

yosefe commented:
/azp run UCX PR
